### PR TITLE
egressgw: fix deadlock due to behavior of RWLock

### PR DESCRIPTION
**Pull request by dcbw**
```   
    1. ensureRouteInfoLocked() locks the cache
    2. getRouteInfosForGateway() tries to grab the cache RLock() but is blocked on 1
    3. ensureRouteInfoLocked() gets the routeInfo from the cache releases the cache lock
    4. getRouteInfosForGateway() acquires the RLock() on the cache and begins iterating routeInfos
    5. ensureRouteInfoLocked() locks the routeInfo
    6. another call to ensureRouteInfoLocked() tries to acquire the cache lock, this blocks all new acquisitions of the cache RLock() per the RWLock documentation
    7. the first ensureRouteInfoLocked() now tries to acquire the cache RLock() to validate that its routeInfo is still in the cache, but is blocked on 6
    8. getRouteInfosForGateway() tries to lock the same routeInfo that 5 has already locked; never releases the cache lock
    
    The core problem is holding the routeInfo lock while trying to acquire the
    route cache lock, which leads to this deadlock. The only reason we try to
    do that is to ensure we don't use a deleted routeInfo object, but we can
    track deletion inside the routeInfo itself instead.
    
    This solution fixes the deadlock but does leave a small window of time
    where, if a gateway pod is created/deleted/created/etc quickly the
    operation will return an error (because the pod's routeInfo is marked
    deleted but not yet removed from the cache) but that's better than
    a deadlock since we can work around it by leaving a bit of time between
    the delete/re-create operations.
```

### CODE REVIEW

**Comment by tssurya** [comment]
so I think the control plane tests are actually timing out:
`Error: The action has timed out.` In https://github.com/ovn-org/ovn-kubernetes/commit/46c13374cc9a4ded776437f447b0b3a1a201a51a we set the time limit to 2.5 hrs and this test took longer than that and got timed out due to which the tests got interrupted.
The external gateway test failed but its because the test pod didn't come up.

File: go-controller/pkg/ovn/egressgw.go [hunk 1]
```diff
@@ -143,23 +141,33 @@ func (oc *Controller) deleteRouteInfoLocked(name ktypes.NamespacedName) *externa
...
+	klog.Warningf("##### [%p %s] deleteRouteInfoLocked() marking routeInfo deleted", routeInfo, name)
+	routeInfo.deleted = true
+
+	go func() {
+		oc.exGWCacheMutex.Lock()
+		defer oc.exGWCacheMutex.Unlock()
+		klog.Warningf("##### [%p %s] deleteRouteInfoLocked() checking routeInfo for removal", routeInfo, name)
+		if newRouteInfo := oc.externalGWCache[name]; routeInfo == newRouteInfo {
+			klog.Warningf("##### [%p %s] deleteRouteInfoLocked() removing routeInfo", routeInfo, name)
+			delete(oc.externalGWCache, name)
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressgw.go:167]: so this makes sure we first check if something changed between acquiring `routeInfo` lock and releasing `exGWCacheMutex` lock. Why did we convert this to a goroutine? It's not guaranteed to finish before we exit out of `deleteRouteInfoLocked` right? So will we get into trouble if this routine delays finishing and we hit `delete(routeInfo.podExternalRoutes, pod)` in `deleteGWRoutesForPod` which could be nil at that point?

File: go-controller/pkg/ovn/egressgw.go [hunk 2]
```diff
@@ -83,15 +86,10 @@ func (oc *Controller) ensureRouteInfoLocked(podName ktypes.NamespacedName) (*ext
...
-	if routeInfoExisted {
-		// Check that the routeInfo wasn't altered in the cache while we were waiting for the lock
-		// and that routeInfo that we locked is not stale.
-		oc.exGWCacheMutex.RLock()
-		defer oc.exGWCacheMutex.RUnlock()
-		if routeInfo != oc.externalGWCache[podName] {
-			routeInfo.Unlock()
-			return nil, fmt.Errorf("routeInfo for pod %s, was altered during ensure route info", podName)
-		}
+	if routeInfo.deleted {
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressgw.go:89]: since we only care about whether it got deleted or not, we should be ok in not wanting to check in any other edits I guess.

File: go-controller/pkg/ovn/egressgw.go [hunk 3]
```diff
@@ -47,7 +47,9 @@ type ovnRoute struct {
 }
 
 type externalRouteInfo struct {
-	sync.RWMutex
+	sync.Mutex
+	deleted bool
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressgw.go:51]: We call `deleteRouteInfoLocked` only from `deleteGWRoutesForPod` which is only called from `deleteLogicalPort`....That is the only path where the pod is entirely removed from the cache. We are sure that we only need to care about the `did it get deleted from the cache` scenario right? I guess we won'tbe in a `did it get re-added` scenario? Since the pod would actually have to be recreated of sorts and since ns/name is unique we should be fine,,

**Review by tssurya — COMMENTED** [review]
Logic wise I think it looks ok, I don't atm see any flaws except for a small question on the goroutine logic inline. Let's maybe give this to the scale team for a few runs and see?

File: go-controller/pkg/ovn/egressgw.go [hunk 4]
```diff
@@ -47,7 +47,9 @@ type ovnRoute struct {
...
+	podName ktypes.NamespacedName
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressgw.go:52]: q: why are we adding this podName here? we have the key on main cache as ns/podName and podExternalRoutes has the pod.IPs as its key. Is it only for logging purposes?

Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:52]: You're right, don't need it here, it was left-over from the cleanup PR.

File: go-controller/pkg/ovn/egressgw.go [hunk 5]
```diff
@@ -389,99 +398,122 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod, namespace string) {
...
-					mask := GetIPFullMask(podIP)
-					node := util.GetWorkerFromGatewayRouter(gr)
-					portPrefix, err := oc.extSwitchPrefix(node)
-					if err != nil {
-						klog.Infof("Failed to find ext switch prefix for %s %v", node, err)
-						continue
-					}
+			klog.Warningf("##### [%p %s] deletePodGWRoutesForNamespace() processing", routeInfo, routeInfo.podName)
+			routeInfo.Lock()
+			if !routeInfo.deleted {
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:369]: Yep.

File: go-controller/pkg/ovn/egressgw.go (hunk 1 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go]: For the goroutine question, the problem is that we cannot attempt to acquire the cache mutex while also holding the routeInfo lock, or we risk the deadlock situation. Punting the actual removal to a goroutine breaks that chain and allows us to remove it slightly later, at the cost of possibly having issues with quick create/delete/create situations; but that's easily worked around by deleting the pod, waiting a couple seconds, and restarting. I'd argue that's preferable to a deadlock, until we convert this to a workqueue.

For the section question about delays, routeInfo is still a reference to the object that we're removing from the cache, and the object is still alive since we're passing the reference around. It just doesn't have an entry in the cache any more after the goroutine is done. But the reference gets passed back to `deleteGWRoutesForPod()`and will be alive until that function exits. So the `delete(routeInfo.podExternalRoutes, pod)` will still be working with a valid routeInfo object.

File: go-controller/pkg/ovn/egressgw.go (hunk 2 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:88]: Yeah, and this part is the tradeoff with no longer deadlocking. If we get to this point, we'll return the error and that means `addGWRoutesForPod()` will fail, and won't add the pod on a quick create/delete/create cycle. The risk is that when the object is deleted we grab the routeInfo lock, set deleted=true, start the goroutine, and return the locked routeInfo, but before the goroutine is actually able to begin executing the pod gets re-added and `ensureRouteInfoLocked()` gets called. It grabs the to-be-deleted routeinfo out of the cache (since the `deleteRouteInfoLocked` goroutine hasn't started yet), locks it, and sees that it's deleted, and returns the error, failing to reach consistent state if no further pod events occur.

But that's better than a deadlock since we can work around it by deleting, waiting a couple seconds, and recreating the pod, IMO.

File: go-controller/pkg/ovn/egressgw.go (hunk 3 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:51]: Our pod events for the same pod are serialized so that greatly lessens this kind of race. But as above, there is a slight chance that a quick create/delete/create could fail the second create, because the `delete` goroutine hasn't run yet and the pod gets added again, and the error is returned.

But, and I just realized this, that makes our pod handler retry! So we're good here anyway, because even if it fails to be updated because it's marked deleted, the pod add handler will retry a bit later and by that point the goroutine will have run, and we'll be OK.

**Comment by dcbw** [comment]
@tssurya I think this is good to go now... I realized that even if we fail to handle the create/delete/create scenario because the goroutine that removes the routeInfo from the cache hasn't quite run yet by the time the ensureRouteInfoLocked for the re-create runs, that just returns an error to addLogicalPort which makes the pod handler retry a bit later, by which point the goroutine will have run. So I think we're OK!

(obviously this is not optimal and we should move to the workqueue model...)

File: go-controller/pkg/ovn/egressgw.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/ovn/egressgw.go:51]: oh yes we will retry! good point. sg

**Review by tssurya — APPROVED** [review]
/lgtm
thanks for explaining each part of the commit Dan, was a complicated change but I think this looks good now and is ready to go. @trozet PTAL

File: go-controller/pkg/ovn/egressgw.go [hunk 6]
```diff
@@ -83,15 +85,9 @@ func (oc *Controller) ensureRouteInfoLocked(podName ktypes.NamespacedName) (*ext
 
 	// 5. If routeInfo originally existed in the cache, there is a chance it was altered after we
```
Comment (trozet) [inline go-controller/pkg/ovn/egressgw.go:86]: this comment is no longer accurate right? We arent checking if it was altered anymore, we are checking if it was deleted.

File: go-controller/pkg/ovn/egressgw.go [hunk 7]
```diff
@@ -390,42 +388,45 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod, namespace string) {
...
-					}
-					mask := GetIPFullMask(podIP)
-					node := util.GetWorkerFromGatewayRouter(gr)
-					portPrefix, err := oc.extSwitchPrefix(node)
-					if err != nil {
-						klog.Infof("Failed to find ext switch prefix for %s %v", node, err)
-						continue
-					}
+			routeInfo.Lock()
+			if !routeInfo.deleted {
```
Comment (trozet) [inline go-controller/pkg/ovn/egressgw.go:369]: how about if routeInfo.deleted { continue } ? save indent on the nested loops

File: go-controller/pkg/ovn/egressgw.go [hunk 8]
```diff
@@ -481,7 +485,7 @@ func (oc *Controller) deleteGWRoutesForPod(name ktypes.NamespacedName, podIPNets
 		if gwToGr, ok := routeInfo.podExternalRoutes[pod]; ok {
 			if len(gwToGr) == 0 {
 				delete(routeInfo.podExternalRoutes, pod)
-				return
+				continue
```
Comment (trozet) [inline go-controller/pkg/ovn/egressgw.go:462]: this looks like a separate bug fix right? Can we move this into its own commit. May also need bz for this?

File: go-controller/pkg/ovn/egressgw.go [hunk 9]
```diff
@@ -130,7 +125,6 @@ func (oc *Controller) getRouteInfosForNamespace(namespace string) []*externalRou
 		if namespacedName.Namespace != namespace {
 			continue
 		}
-		routeInfo.Lock()
```
Comment (trozet) [inline go-controller/pkg/ovn/egressgw.go:133]: comment on these functions are no longer accurate "locked for a specific namespace"

File: go-controller/pkg/ovn/egressgw.go [hunk 10]
```diff
@@ -93,39 +93,16 @@ func (oc *Controller) ensureRouteInfoLocked(podName ktypes.NamespacedName) (*ext
 	return routeInfo, nil
 }
 
-// getRouteInfosForGateway returns all routeInfos locked for a specific namespace and gateway IP
-func (oc *Controller) getRouteInfosForGateway(gatewayIP, namespace string) []*externalRouteInfo {
```
Comment (trozet) [inline go-controller/pkg/ovn/egressgw.go:101]: The commit message seems confusion to me. This part:
"getRouteInfosForGateway() accesses a routeInfo's podExternalRoutes
without taking the routeInfo's lock."

with your previous commit it removes locking routeInfo from this function as well as getRouteInfosForNamespace. So I'm not sure how this is part of the reasoning to remove the function.

File: go-controller/pkg/ovn/egressgw.go [hunk 11]
```diff
@@ -386,7 +363,7 @@ func (oc *Controller) deletePodGWRoutesForNamespace(pod, namespace string) {
 
 	for _, gwIP := range foundGws.gws {
 		// check for previously configured pod routes
-		routeInfos := oc.getRouteInfosForGateway(gwIP.String(), namespace)
+		routeInfos := oc.getRouteInfosForNamespace(namespace)
```
Comment (trozet) [inline go-controller/pkg/ovn/egressgw.go:366]: the downside to this is you are going to lock more routeInfos unnecessarily. I don't see much value in this 2nd commit.

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/egressgw.go (hunk 8 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:462]: done

File: go-controller/pkg/ovn/egressgw.go (hunk 7 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:369]: If you're cool with:

```
routeInfo.Lock()
if routeInfo.deleted {
    routeInfo.Unlock()
    continue
}
...
routeInfo.Unlock()
```

File: go-controller/pkg/ovn/egressgw.go (hunk 9 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:133]: done

File: go-controller/pkg/ovn/egressgw.go (hunk 10 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:101]: The original code did this:

```
 		for _, route := range routeInfo.podExternalRoutes {
 			if _, ok := route[gatewayIP]; ok {
				routeInfo.Lock()
 				routeInfos = append(routeInfos, routeInfo)
 				break
 			}
```

it accessed `routeInfo.podExternalRoutes` and *then* took the routeInfo lock. But the caller of this function already does the `route[gatewayIP]` check anyway, so it's pointless to have this function now...

File: go-controller/pkg/ovn/egressgw.go (hunk 11 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:366]: It fixes a real concurrent data access in the original code. There's no way to safely filter the routeInfos how `getRouteInfosForGateway()` was doing without locking each and every routeInfo in that namespace, right? You have to lock to check whether podExternalRoutes[] has the gatewayIP you care about.

File: go-controller/pkg/ovn/egressgw.go (hunk 6 above)
Comment (dcbw) [reply go-controller/pkg/ovn/egressgw.go:86]: good point, fixed.

File: go-controller/pkg/ovn/egressgw.go (hunk 11 above)
Comment (trozet) [reply go-controller/pkg/ovn/egressgw.go:371]: ah now it makes sense. Thanks.

**Review by trozet — APPROVED** [review]
